keps/sig-api-machinery/20190603-immutable-fields.md

```diff
@@ -46,100 +56,112 @@ i.e. after creation of an object, those fields cannot be mutated anymore. E.g. a
-could potentially add that semantics later.
-
-It's also important to note that this concept is orthogonal to field ownership
-and list associativity. Ideally we would manage to keep the concepts entirely
-separate. e.g. An atomic list is a different concept from an immutable list.
-
-There are different semantics for a field to be immutable, especially if we
-consider non-leaf fields like lists and maps.
-
-#### Field selection
-
-Scalar fields have an obvious "selection" mechanism: they are either immutable
-or not. Things get more complicated for lists and maps, since we need to know if
-the flag applies to the list/map itself, to its members, or both.
-
-A few possible semantics are described here:
+We propose
+
+1. adding a vendor extensions to CRD OpenAPI schemas named `x-kubernetes-immutable` with boolean
+   value, but only `true` is a valid value.
+2. do **strict deep-equal** comparison of those fields listed in this extension during
+   update validation in the request pipeline. If that comparison fails, the request is 
+   rejected as 400 "Bad Request".
+   
+We create another KEP to define custom normalization steps for CRs done during 
+deserialization from etcd and when receiving a request (just after pruning and defaulting).
+
+### OpenAPI extension
+
+The `x-kubernetes-immutable` vendor extension is set in the spec of fields,
+arrays and objects. It recursively applies to the whole subtree:
+
+```yaml
+properties:
+  foo:
+    type: array
+    items:
+      x-kubernetes-immutable: true
+      type: string
+```
 
-- Non-recursive: For a list or a map, one can not remove or add new items, but
-  existing items can be modified.
-- Recursive: means that none of their field can change, and new fields are not
-  accepted.
-- Recursive with addition and/or deletion:
-  - New/deleted fields in maps and lists are tolerated.
-  - The added/deleted type would probably be "Recursive" after the first level
-    (addition/delete is not recursive).
-  - For associative lists, keys are already "immutable" in a way (similar to how
-    names are "immutables" for objects). Changing the keys would mean deleting
-    and creating a new item in the list, which would be permitted by these
-    rules.
-- Mutable: A field set as mutable could cancel the recursive immutability.
+means that the `foo` array itself is mutable, but the items are not. This means that
+items can be deleted and added, but at each existing index the string values are
+immutable.
 
-#### On mutation of selected fields
+```yaml
+properties:
+  foo:
+    type: object
+    x-kubernetes-immutable: true
+```
 
-- Error: mutation of an immutable field returns an error and the request is
-  rejected. No mutation is performed at all.
-- Ignored: mutation of the field (and/or sub-fields) are completely ignored, and
-  not persisted. They do not trigger an error.
+means that the whole `.foo` object can be removed if set, or set if it is unset, but
```

Comment:
allowing a delete and a recreate with different values defies my expectations.  This is consistent with how it's treated for objects in some way? Seems weird to allow removing entire stanzas to remove values.  Is this to make it work nicely with unions?

Reply:
I *think* this is what @deads2k is already alluding to, but how is a delete and insert identified differently from an update of an entry in-place? It seems like this sort of thing would only be possible if using a list that is keyed on, e.g. a `name` field. Even then, I agree it seems like an odd thing to permit as from a users perspective, they now just have to perform two API operations instead of one in order to update certain fields.

Reply:
I get your point. tl/dr: you question the sense of pure removal/addition without identifier. I don't object.

What about string slices? Is the identifier implicitly the value and we would allow a deletion and an insertion in a special mode, but no order changes?

Note, we are very flexible what to allow. The big question is what we really want to offer in Kubernetes-like APIs defined through CRDs.

Reply:
Not sure your two question are actually the same. I fixed what I think @deads2k pointed out. My example was wrong. Specifying `x-kubernetes-immutable` applies to the field itself, not only the sub-fields in case of an object.